File: backend/src/gtin.ts
// Mirror of scripts/fdc/gtin.py. Canonical form: 13-digit zero-padded GTIN with check digit.
export function normalizeGTIN(raw: string | null | undefined): string | null {
  if (!raw) return null;
  const compact = raw.replace(/[\s-]/g, "");
  if (!/^\d+$/.test(compact)) return null;

  let canonical: string;
  if (compact.length === 8) canonical = "00000" + compact;
  else if (compact.length === 12) canonical = "0" + compact;
  else if (compact.length === 13) canonical = compact;
  else if (compact.length === 14 && compact.startsWith("0")) canonical = compact.slice(1);
  else return null;

  const digits = [...canonical].map(Number);
  const supplied = digits[12];
  const sum = digits.slice(0, 12).reverse().reduce(
    (total, digit, index) => total + digit * (index % 2 === 0 ? 3 : 1),
    0,
  );
  return (10 - (sum % 10)) % 10 === supplied ? canonical : null;
}

File: scripts/fdc/gtin.py
"""Canonical barcode form: 13-digit zero-padded GTIN with check digit.

Mirror of backend/src/gtin.ts and ScannerViewModel.canonicalBarcode.
"""
import re

_SEPARATORS = re.compile(r"[\s-]")


def normalize_gtin(raw: str) -> str | None:
    if not raw:
        return None
    compact = _SEPARATORS.sub("", raw)
    if not compact.isascii() or not compact.isdigit():
        return None

    if len(compact) == 8:
        canonical = "00000" + compact
    elif len(compact) == 12:
        canonical = "0" + compact
    elif len(compact) == 13:
        canonical = compact
    elif len(compact) == 14 and compact.startswith("0"):
        canonical = compact[1:]
    else:
        return None

    digits = [int(value) for value in canonical]
    total = sum(value * (3 if index % 2 == 0 else 1)
                for index, value in enumerate(reversed(digits[:-1])))
    expected = (10 - total % 10) % 10
    return canonical if expected == digits[-1] else None
